File: Java/net.alcetech.ApplicationFramework.Core/src/net/alcetech/ApplicationFramework/Core/Application.java
package net.alcetech.ApplicationFramework.Core;

import java.util.ArrayList;
import java.util.Hashtable;

public class Application {
	private static String[] _args = new String[0];
	public static void initialize(String[] args) {
		Application._args = args;
	}
	
	private static String _Title = "";
	public static String getTitle() { return _Title; }
	public static void setTitle(String value) { _Title = value; }

	private static ArrayList<Command> _commands;
	public static void addCommand(Command item) {
		_commands.add(item);
	}
	public static Command getCommand(String name) {
		for (int i = 0; i < _commands.size(); i++) {
			Command item = _commands.get(i);
			if (item.getName().equals(name)) return item;
		}
		return null;
	}
	
	public static boolean executeCommand(Command item) {
		if (item == null) return false;
	
		if (item.getID().equals("FileExit"))
		{
			System.exit(0);
			return true;
		}
		
		return false;
	}

	
	private static ArrayList<CommandItem> _mainMenuCommandItems;
	public static CommandItem[] getMainMenuCommandItems() {
		CommandItem[] array = new CommandItem[_mainMenuCommandItems.size()];
		_mainMenuCommandItems.toArray(array);
		return array;
	}
	public static void addMainMenuCommandItem(CommandItem item) {
		_mainMenuCommandItems.add(item);
	}
	
	private static Hashtable<Command, ArrayList<ICommandListener>> _commandBindings = new Hashtable<Command, ArrayList<ICommandListener>>();
	public static void bindCommand(Command item, ICommandListener listener) {
		if (item == null) {
			System.out.println("appfx: Application::bindCommand() - item is null\n");
			return;
		}
		
		if (!_commandBindings.containsKey(item))
		{
			_commandBindings.put(item, new ArrayList<ICommandListener>());
		}
		
		ArrayList<ICommandListener> bindings = _commandBindings.get(item);
		bindings.add(listener);
	}
	
	public static void start() {
		
	}
	public static void stop() {
		stop(0);
	}
	public static void stop(int exitCode) {
		System.exit(exitCode);
	}

}

File: CPlusPlus/net.alcetech.ApplicationFramework.Core/Application.cpp
/*
 * Application.cpp
 *
 *  Created on: Apr 6, 2016
 *      Author: beckermj
 */

#include "Application.h"

namespace ApplicationFramework {

List<Command*>* Application::_commands;
List<CommandItem*>* Application::_mainMenuCommandItems;
Dictionary<Command*, List<void(*)(Command*)>*>* Application::_commandBindings;

int Application::_argc;
char** Application::_argv;

String* Application::_title = String::EMPTY;

void Application::initialize(int argc, char** argv) {
	Application::_argc = argc;
	Application::_argv = argv;

	Application::_commands = new List<Command*>();
	Application::_mainMenuCommandItems = new List<CommandItem*>();
	Application::_commandBindings = new Dictionary<Command*, List<void(*)(Command*)>*>();
}

void Application::addCommand(Command* item) {
	Application::_commands->add(item);
}
Command* Application::getCommand(const char* name) {
	for (int i = 0; i < Application::_commands->count(); i++)
	{
		Command* item = Application::_commands->get(i);
		if (item->getID()->equals(name)) return item;
	}
	return NULL;
}
Command* Application::getCommand(String* name) {
	return Application::getCommand(name->toCharArray());
}

bool Application::executeCommand(Command* item) {
	if (item == NULL) return false;

	if (item->getID()->equals("FileExit"))
	{
		exit(0);
		return true;
	}

	List<void(*)(Command*)>* list = Application::_commandBindings->get(item);
	if (list != NULL)
	{
		for (int i = 0; i < list->count(); i++)
		{
			list->get(i)(item);
		}
		return true;
	}
	return false;
}

void Application::addMainMenuCommandItem(CommandItem* item) {
	Application::_mainMenuCommandItems->add(item);
}
List<CommandItem*>* Application::getMainMenuCommandItemsList() {
	return Application::_mainMenuCommandItems;
}

void Application::bindCommand(Command* item, void (*callback)(Command* command)) {
	if (item == NULL) {
		printf("Application::bindCommand() - item is null\n");
		return;
	}
	List<void(*)(Command*)>* bindings = Application::_commandBindings->get(item);
	if (bindings == NULL)
	{
		bindings = new List<void(*)(Command*)>();

		Application::_commandBindings->add(item, bindings);
	}
	bindings->add(callback);
}

String* Application::getTitle() {
	return Application::_title;
}
void Application::setTitle(String* value) {
	Application::_title = value;
}

} /* namespace ApplicationFramework */
